sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/AlterTableExec.scala

```diff
@@ -44,3 +46,35 @@ case class AlterTableExec(
     Seq.empty
   }
 }
+
+/**
+ * Physical plan node for adding a check constraint with validation.
+ */
+case class AddCheckConstraintExec(
+    catalog: TableCatalog,
+    ident: Identifier,
+    change: TableChange,
+    condition: String,
+    child: SparkPlan) extends V2CommandExec with UnaryExecNode {
+
+  override def output: Seq[Attribute] = Seq.empty
+
+  override protected def run(): Seq[InternalRow] = {
+    if (child.executeTake(1).nonEmpty) {
```

Comment:
Hmm, I'm not sure if this question makes sense, just wondering if it could be a problem.

As we don't lock the table access, I assume that any concurrent change could happen during altering the table with a new check constraint. If so, isn't it possible that some data violating the new check constraint could be added after this validation?

Reply:
The check constraint is on specified table version. But seems not all tables support versioning.

Reply:
This is something that would be handled in the connector. Spark is not aware of the concurrent changes as they could happen in other engines. That's why it is important to pass the version of the table that was validated back to the connector. The connector can then check whether there were new operations that violate the constraint.

Reply:
A potential solution for this:
* The current table version is pinned after loading the table
* The current table version is not used for plan comparison when refreshing cache.
* if the table supports versions and time travel, scan the table with time travel; otherwise, simply scan the table as it is in this PR since the validated table version will be null anyway.